graph/src/extension.rs

```diff
@@ -0,0 +1,120 @@
+use futures::sync::oneshot;
+use std::sync::{Arc, Mutex};
+use tokio::prelude::{Future, Poll, Stream};
+
+pub struct Cancelable<T, C> {
+    cancelable: T,
+    canceled: oneshot::Receiver<()>,
+    on_cancel: C,
+}
+
+/// It's not viable to use `select` directly, so we do a custom implementation.
+impl<S: Stream, C: Fn() -> S::Error> Stream for Cancelable<S, C> {
+    type Item = S::Item;
+    type Error = S::Error;
+
+    fn poll(&mut self) -> Poll<Option<Self::Item>, Self::Error> {
+        // Error if the stream was canceled, otherwise poll it.
+        if self.canceled.poll().is_err() {
+            Err((self.on_cancel)())
+        } else {
+            self.cancelable.poll()
+        }
+    }
+}
+
+impl<F: Future, C: Fn() -> F::Error> Future for Cancelable<F, C> {
+    type Item = F::Item;
+    type Error = F::Error;
+
+    fn poll(&mut self) -> Poll<Self::Item, Self::Error> {
+        // Error if the future was canceled, otherwise poll it.
+        if self.canceled.poll().is_err() {
+            Err((self.on_cancel)())
+        } else {
+            self.cancelable.poll()
+        }
+    }
+}
+
+#[derive(Debug)]
+pub struct CancelGuard {
+    canceler: oneshot::Sender<()>,
+}
+
+impl CancelGuard {
+    /// A more readable `drop`.
+    pub fn cancel(self) {}
+
+    /// Convert into a `SharedGuard`.
+    pub fn shared(self) -> SharedCancelGuard {
+        SharedCancelGuard {
+            guard: Arc::new(Mutex::new(Some(self.canceler))),
+        }
+    }
+}
+
+/// A version of `CancelGuard` that can be cloned.
+///
+/// To cancel the stream, call `cancel` on one of the guards or drop all guards.
+#[derive(Clone, Debug)]
+pub struct SharedCancelGuard {
+    guard: Arc<Mutex<Option<oneshot::Sender<()>>>>,
+}
+
+impl SharedCancelGuard {
+    /// Cancels the stream, a noop if already canceled.
+    pub fn cancel(&self) {
+        *self.guard.lock().unwrap() = None
+    }
+
+    pub fn is_canceled(&self) -> bool {
+        self.guard.lock().unwrap().is_some()
+    }
+}
+
+pub trait StreamExtension: Stream + Sized {
+    /// When `cancel` is called on a `CancelGuard` or it is dropped,
+    /// `Cancelable` receives an error.
+    ///
+    /// `on_cancel` is called to make an error value upon cancelation.
+    fn cancelable<C: Fn() -> Self::Error>(self, on_cancel: C)
```

Comment:
@timmclean How do you feel about the `on_cancel` callback? Is that something that would work well with cascading streams?

I'm not sure if I'd prefer an equivalent to `map_sink_err` (e.g. `map_cancelable_error`) instead of passing the callback in to this function. It feels slightly more idiomatic but also less obvious. @leodasvacas What do you think?

Reply:
The error handling is tricky here. `Cancelable` supports the `map_err` methods since it's a stream or future, the problem is determining what should be the error type of `Cancelable`, which must be convertible from `Self::Error`.

@timmclean suggested an enum approach in the issue, but I think overly nested errors can get hard to manage. My first idea was requiring `Self::Error: From<Canceled>` but it was hard to use.

The closure I think is pretty flexible and ergonomic, for example it allows `Self::Error` to be an enum with a `Canceled` variant. Though it doesn't resemble the standard futures and stream APIs. It's my best idea so far, let's see how it fares as we start to use this more.